core/src/com/unciv/ui/CivilopediaScreen.kt

```diff
@@ -14,84 +16,202 @@ import com.unciv.ui.tilegroups.TileGroup
+            }
+        }
+
+        description.clearListeners()
+
+        val selectedEntry = categoryMap.entries[selectEntry]
+        if (selectedEntry == null) {
+            description.setText("")
+        } else {
+            selectedEntry.button?.let {
+                println("Civilopedia select ${category.label}.$selectEntry: (${it.x},${it.y},${it.width},${it.height}), scroll.y=${entrySelectScroll.y}, table.y/h=${entrySelectTable.y},${entrySelectTable.height}")
+                entrySelectScroll.scrollTo(it.x, it.y, it.width, it.height, true, false)
+                it.highlight()
+            }
+            description.setText(selectedEntry.description)
+
+            //// TESTING CODE
+            // right now, the description is just a big text label, we'd need to split that to make
+            // portions clickable individually...
+            var targetCategory: CivilopediaCategories = category
+            var targetEntry: String? = null
+            if (category == CivilopediaCategories.Units) {
+                val unit = ruleset.units[selectEntry]
+                when {
+                    unit?.uniqueTo != null -> { targetCategory = CivilopediaCategories.Nations; targetEntry = unit.uniqueTo }
+                    unit?.upgradesTo != null -> { targetEntry = unit.upgradesTo }
+                }
+            }
+            if (category == CivilopediaCategories.Buildings) {
+                val building = ruleset.buildings[selectEntry]
+                when {
+                    building?.uniqueTo != null -> { targetCategory = CivilopediaCategories.Nations; targetEntry = building.uniqueTo }
+                    building?.requiredTech != null -> { targetCategory = CivilopediaCategories.Technologies; targetEntry = building.requiredTech }
+                }
+            }
+            if (targetCategory != category || targetEntry != null) {
+                description.onClick { select(targetCategory, targetEntry) }
+            }
+        }
+
+    }
+
+    private fun Button.highlight (highlight: Boolean = true) {
+        if (highlight) {
+            highlightedButton?.highlight(false)
+            highlightedButton = null
         }
+        val newColor = if (highlight) Color.GOLDENROD else Color.WHITE
+        //(this.children.firstOrNull { it is Image } as Image?)?.color = newColor
+        (this.children.firstOrNull { it is Label } as Label?)?.color = newColor
+        if (highlight)
+            highlightedButton = this
     }
 
     init {
         onBackButtonClicked { UncivGame.Current.setWorldScreen() }
 
+        categoryMaps[CivilopediaCategories.Buildings] = CivilopediaCategoryMap().apply {
+            entries.putAll (
+                ruleset.buildings.map {
+                    Pair (it.key, CivilopediaEntry(it.value.getDescription(false, null,ruleset),
+                        ImageGetter.getConstructionImage(it.key)))
+                }.sortedBy { it.first }
+            )
+        }
```

Comment:
These are in desperate need of refactoring - if you've taken what used to be  lines and ended up with 6, something's probably wrong.
Each value maps to a CivilopediaEntry in the old code, right? And the CivilopediaCategoryMap contains a bunch of those? Then it should receive a collection of them as a construction parameter, and bam - we're back to 3, maybe 4 lines each, with all the boilerplate happening in the CivilopediaCategoryMap class

Reply:
Did I already say I think all these 'game object' classes need a common interface? One supporting a getDescription and getImage?

If that existed this here would have been easier and cleaner. As is, all those getDescription calls have incompatible signatures and operate on incompatible classes. Worse for the images. That was a task I was too chicken to work on, thus the ugliness. It did each category separately before, already ugly, and with a common interface that would be one nested loop.